### ISSUE TITLE
Questions about advanced ECS features in comparison to Unity ECS

### ISSUE THREAD

**Issue by drywolf** [SanderMertens/flecs#162]
I have experimented with the Unity ECS / DOTS previously and am now looking for a comparable solution for usage in C/C++ hence I am comparing it to flecs.

I very much like the API design of flecs so far and I can see that it shares some of the concepts with the Unity ECS architecture but other things are solved very much differently.
Some of the features that the Unity ECS provides I couldn't figure out how to achieve or if those are fundamentally not yet implemented in flecs:

* [Shared component data](https://docs.unity3d.com/Packages/com.unity.entities@0.7/manual/shared_component_data.html) allows to separate entities into "categories" (dedicated chunks) based on the combinatorial values of a specially designated component-type `ISharedComponentData`

* [Dynamic buffer components](https://docs.unity3d.com/Packages/com.unity.entities@0.7/manual/dynamic_buffers.html) allow to associate array-like data with an entity. The array-elements are stored within a chunk up to a fixed reserved size, if this number of elements is exceeded for an entity then the array will be moved out of the chunk into heap memory.

* [Chunk component data](https://docs.unity3d.com/Packages/com.unity.entities@0.7/manual/ecs_chunk_component.html) can be used to store meta-data that is common for all entities contained within a chunk (for example an AABB for all entities within a chunk ... as shown in the unity docs)

* [Chunk version numbers](https://docs.unity3d.com/Packages/com.unity.entities@0.7/manual/version_numbers.html) ... In the Unity ECS, system access to components always has to define wether readonly access or write access is required by the system code. If component-data in a chunk is accessed by a system with write-access, then this increments the ChangeVersion stored within the chunk for this component type. By doing this the ECS can efficiently implement a coarse change tracking mechanism, i.e. it is possible to run a system / jobs only for a set of entities that have (potentially) changed values compared to the previous frame. In other words, if no system wrote to a certain component-type since the last frame and another system only requests readonly access for this component-type AND the used [EntityQuery has a change-filter](https://docs.unity3d.com/Packages/com.unity.entities@0.7/manual/ecs_entity_query.html#change-filters) enabled, then the systen will not have to process any of those entities.

**TLDR:**
* would it be possible with the current flecs API already to achive the above features or should I look into implementing those alongside the existing features ?
* does flecs underlying storage use an equivalent concept of `Chunks` as Unity describes it in their documents ? Is there documentation about how flecs organizes its low-level storage ? I've seen the `ecs_chunked_t` type in the code but I don't have a complete enough understanding of the code to know if this is an equivalent to what Unity has implemented.

@SanderMertens Thanks for putting up with my wall of questions 😉 
I really like flecs so far and just want to dig deeper what could be possible.

**Comment by SanderMertens** [comment]
@drywolf great questions! Let me try to answer them:

**1. Shared components**
Shared components are supported in Flecs, but in a different way. In Flecs you share components _between entities_, and the feature is called inheritance: https://github.com/SanderMertens/flecs/blob/master/Manual.md#inheritance

By adding an inheritance relationship to an entity, you change its archetype which means that it fragments the data space, which means that the end result is similar to what happens in Unity. For example, if I have 100 entities with the same components, but 0-49 share from BaseA and 50-99 share from BaseB, they will be stored in separate archetypes (arrays), which is (very) roughly equivalent to a chunk.

I recently wrote a blog that gives a bit more background on how inheritance ("prefabs") and hierarchies impact entity types: https://medium.com/@ajmmertens/building-an-ecs-1-types-hierarchies-and-prefabs-9f07666a1e9d

**2. Dynamic buffer components**
Currently you can do this in Flecs by leveraging the OnAdd and OnRemove handlers in a way that is described here: https://github.com/SanderMertens/flecs/blob/master/Manual.md#memory-management. Near-future versions of Flecs will address this more elegantly, with per-component callbacks that mimic C++ operations (copy, move, ctor, dtor).

**3. Chunk component data **
This can be addressed using the same inheritance mechanism.

**4. Chunk version numbers**
There isn't really an equivalent of this in Flecs, however you can annotate components in system signatures with in, inout and out. Currently this is used to force const-ness of components in C++ and to enable "on demand systems" (systems that are only ran if there's a demand for its outputs) but a feature that lets you track changed entities could be implemented.

**5. Does Flecs use chunks **
Today Flecs does not use chunks in the same way that Unity does. Components are stored in using SoA (strucs of arrays) and there is one array per component per archetype. Future versions of Flecs may support chunks as it has some advantages (no reallocs, easier thread scheduling) but that is likely not going to happen soon.

The `ecs_chunked_t` datatype is actually a sparse set (it has been renamed in future branches) which happens to store data in chunks. Some internal data structures use this so pointers remain constant, but that's all hidden for the user. One area in particular where this data structure is going to be used heavily is for the entity index (see the blog I linked to).

Hope that helps! Feel free to join Discord if you'd like to discuss some of these things in more detail (though I'm also more than happy to answer questions on GH).

**Comment by SanderMertens** [comment]
Closing this issue, feel free to reopen if you have more questions.

**Comment by drywolf** [comment]
Thanks @SanderMertens for the detailed explanations. It really helped a lot.
I'll make sure to join Discord.

Thanks for this great project 👍
